### PR TITLE
 Add Annotation field to reports and batches

### PR DESCRIPTION
**Pull request by dcreager**
This lets you write custom processors that annotate the reports with arbitrary additional information.  The test case is a simple example of looking up geo information for the client, and serving infrastructure information for the server.

### CODE REVIEW

File: pkg/collector/pipeline_test.go [hunk 1]
```diff
@@ -121,3 +123,89 @@ func TestDumpReports(t *testing.T) {
...
+type geoAnnotator struct{}
+
+func (g geoAnnotator) ProcessReports(batch *ReportBatch) {
+	batch.Annotation = clientCountries[batch.ClientIP]
+	for i, _ := range batch.Reports {
+		batch.Reports[i].Annotation = serverZones[batch.Reports[i].ServerIP]
+	}
+}
+
+type stashReports struct {
```
Comment (sburnett) [inline pkg/collector/pipeline_test.go]: Thinking aloud: Could you get away with: `type stashReports = *ReportBatch`?

File: pkg/collector/pipeline_test.go [hunk 2]
```diff
@@ -121,3 +123,89 @@ func TestDumpReports(t *testing.T) {
...
+	request.Header.Add("Content-Type", "application/report")
+	var response httptest.ResponseRecorder
+	pipeline.ServeHTTP(&response, request)
+
+	if response.Code != http.StatusNoContent {
+		t.Errorf("ServeHTTP(%s): got %d, wanted %d", compactJSON(json), response.Code, http.StatusNoContent)
+		return
+	}
+
+	want := ReportBatch{
```
Comment (sburnett) [inline pkg/collector/pipeline_test.go]: It feels a little weird to have the input in a file and the expected output hard coded here. Would it be appropriate to marshal the output back into a file so you can have a golden output? Or would that be too much extra functionality for the test?

File: pkg/collector/pipeline_test.go (hunk 1 above)
Comment (dcreager) [reply pkg/collector/pipeline_test.go]: I tried that; you can't attach a method to a pointer type.

File: pkg/collector/pipeline_test.go (hunk 2 above)
Comment (dcreager) [reply pkg/collector/pipeline_test.go]: I like that idea!  Done.

**Pull request merged**